tests/api/utils/test_definition_rendering.py

```diff
@@ -0,0 +1,303 @@
+                "source_stage": "stage <% ctx.native_app.name %>",
+            },
+        },
+    ],
+)
+def test_resolve_variables_error_on_cycle(definition):
+    with pytest.raises(ClickException) as err:
+        render_definition_template(definition)
+
+    assert err.value.message == f"Cycle detected"
+
+
+@pytest.mark.parametrize(
+    "definition, error_var",
+    [
+        (
+            {
+                "definition_version": "1.1",
+                "native_app": {
+                    "name": "app_name",
+                    "artifacts": [{"src": "src/add.py", "dest": "add.py"}],
+                },
+                "env": {"A": "<% ctx.native_app.artifacts %>"},
+            },
+            "ctx.native_app.artifacts",
+        ),
+        (
+            {
+                "definition_version": "1.1",
+                "native_app": {
+                    "name": "app_name",
+                    "artifacts": [{"src": "src/add.py", "dest": "add.py"}],
+                },
+                "env": {"A": "<% ctx.native_app %>"},
+            },
+            "ctx.native_app",
+        ),
+    ],
+)
+def test_resolve_variables_reference_non_scalar(definition, error_var):
+    with pytest.raises(UndefinedError) as err:
+        render_definition_template(definition)
+
+    assert (
+        err.value.message
+        == f"Template variable {error_var} does not contain a valid value"
+    )
+
+
+@mock.patch.dict(os.environ, {"blank_env": ""}, clear=True)
+def test_resolve_variables_blank_is_ok():
+    definition = {
+        "definition_version": "1.1",
+        "native_app": {
+            "name": "<% ctx.env.blank_default_env %>",
+            "source_stage": "",
+            "deploy_root": "<% ctx.env.blank_env %>",
+        },
+        "env": {
+            "blank_default_env": "",
+            "refers_to_blank": "<% ctx.native_app.source_stage %>",
+        },
+    }
+    result = render_definition_template(definition)
+
+    assert result == {
+        "definition_version": "1.1",
+        "native_app": {"name": "", "source_stage": "", "deploy_root": ""},
+        "env": {
+            "blank_env": "",
+            "blank_default_env": "",
+            "refers_to_blank": "",
+        },
+    }
+
+
+@pytest.mark.parametrize(
+    "env, msg",
+    [
+        ({"app": "<% bdbdbd %>"}, "Could not find template variable bdbdbd"),
+        (
+            {"app": "<% ctx.streamlit.name %>"},
+            "Could not find template variable ctx.streamlit.name",
+        ),
+        ({"app": "<% ctx.foo %>"}, "Could not find template variable ctx.foo"),
+    ],
+)
+def test_resolve_variables_fails_if_referencing_unknown_variable(env, msg):
+    definition = {
+        "definition_version": "1.1",
+        "env": env,
+    }
+    with pytest.raises(UndefinedError) as err:
+        render_definition_template(definition)
+    assert msg in str(err.value)
+
+
+@mock.patch.dict(os.environ, {}, clear=True)
```

Comment:
can we also add a test that clearly documents what happens when one tries to inject complex YAML through a variable? e.g. can I add artifact mappings through env vars, etc? We either allow it or we don't, but it should be documented as it is part of the implied contract here.

Reply:
added a test. We just treat the env as a string. We never yaml expand it.